keras_hub/src/models/qwen_moe/qwen_moe_decoder.py

```diff
@@ -0,0 +1,500 @@
+        shared_expert_intermediate_dim,
+        num_experts,
+        top_k,
+        norm_topk_prob,
+        kernel_initializer="glorot_uniform",
+        layer_norm_epsilon=1e-5,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.hidden_dim = hidden_dim
+        self.moe_intermediate_dim = moe_intermediate_dim
+        self.shared_expert_intermediate_dim = shared_expert_intermediate_dim
+        self.num_experts = num_experts
+        self.top_k = top_k
+        self.norm_topk_prob = norm_topk_prob
+        self.kernel_initializer = kernel_initializer
+        self.layer_norm_epsilon = layer_norm_epsilon
+
+    def build(self, decoder_sequence_shape):
+        self._sparse_feedforward_gate_dense = keras.layers.Dense(
+            self.num_experts,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            use_bias=False,
+            dtype=self.dtype_policy,
+            name="sparse_feedforward_gate_dense",
+        )
+        self._sparse_feedforward_gate_dense.build(decoder_sequence_shape)
+
+        self.experts = [
+            QwenMoeMLP(
+                intermediate_dim=self.moe_intermediate_dim,
+                hidden_dim=self.hidden_dim,
+                kernel_initializer=self.kernel_initializer,
+                layer_norm_epsilon=self.layer_norm_epsilon,
+            )
+            for _ in range(self.num_experts)
+        ]
+        for expert in self.experts:
+            expert.build(decoder_sequence_shape)
+
+        self.shared_expert_dense = QwenMoeMLP(
+            intermediate_dim=self.shared_expert_intermediate_dim,
+            hidden_dim=self.hidden_dim,
+            kernel_initializer=self.kernel_initializer,
+            layer_norm_epsilon=self.layer_norm_epsilon,
+        )
+        self.shared_expert_dense.build(decoder_sequence_shape)
+
+        self.shared_expert_gate_dense = keras.layers.Dense(1, use_bias=False)
+        self.shared_expert_gate_dense.build(decoder_sequence_shape)
+        self.built = True
+
+    def call(self, hidden_states):
+        batch_size, seq_len, hidden_dim = hidden_states.shape
+        hidden_states = hidden_states.reshape(-1, hidden_dim)
+
+        router_logits = self._sparse_feedforward_gate_dense(hidden_states)
+
+        routing_weights = ops.softmax(router_logits, axis=1)
+        routing_weights, selected_experts = ops.top_k(
+            routing_weights, self.top_k
+        )
+
+        if self.norm_topk_prob:
+            routing_weights /= ops.sum(routing_weights, axis=-1, keepdims=True)
+
+        routing_weights = ops.cast(routing_weights, hidden_states.dtype)
+
+        final_hidden_states = ops.zeros(
+            (batch_size * seq_len, hidden_dim), dtype=hidden_states.dtype
+        )
+
+        expert_mask = ops.one_hot(
+            selected_experts, num_classes=self.num_experts
+        )
+        expert_mask = ops.transpose(expert_mask, axes=[2, 1, 0])
+
+        for expert_idx in range(self.num_experts):
```

Comment:
The for loop here to go over the experts are inefficient for XLA compilation. This implementation would need to updated - I had tried out a dummy MOE implementation in JAX here - https://colab.sandbox.google.com/drive/1r0rscZK_2bNpDmFLC1POEEQoKcqWQYlQ
I order to bring this to KearsHub we are missing ragged_dot op.

Reply:
I have prototyped the implementation - will add this op soon

Reply:
When can we expect this to be available as a part of keras op?